File: install/include/ql/time/daycounters/actual365fixed.cpp
/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */

/*
 Copyright (C) 2000, 2001, 2002, 2003 RiskMap srl
 Copyright (C) 2013 BGC Partners L.P.

 This file is part of QuantLib, a free-software/open-source library
 for financial quantitative analysts and developers - http://quantlib.org/

 QuantLib is free software: you can redistribute it and/or modify it
 under the terms of the QuantLib license.  You should have received a
 copy of the license along with this program; if not, please email
 <quantlib-dev@lists.sf.net>. The license is also available online at
 <http://quantlib.org/license.shtml>.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the license for more details.
*/

#include <ql/time/daycounters/actual365fixed.hpp>
#include <cmath>

namespace QuantLib {

    ext::shared_ptr<DayCounter::Impl>
    Actual365Fixed::implementation(Actual365Fixed::Convention c) {
        switch (c) {
          case Standard:
            return ext::shared_ptr<DayCounter::Impl>(new Impl);
          case Canadian:
            return ext::shared_ptr<DayCounter::Impl>(new CA_Impl);
          case NoLeap:
            return ext::shared_ptr<DayCounter::Impl>(new NL_Impl);
          default:
            QL_FAIL("unknown Actual/365 (Fixed) convention");
        }
    }

    Time Actual365Fixed::CA_Impl::yearFraction(const Date& d1,
                                               const Date& d2,
                                               const Date& refPeriodStart,
                                               const Date& refPeriodEnd) const {
        if (d1 == d2)
            return 0.0;

        // We need the period to calculate the frequency
        QL_REQUIRE(refPeriodStart != Date(), "invalid refPeriodStart");
        QL_REQUIRE(refPeriodEnd != Date(), "invalid refPeriodEnd");

        Time dcs = daysBetween(d1,d2);
        Time dcc = daysBetween(refPeriodStart,refPeriodEnd);
        auto months = Integer(std::lround(12 * dcc / 365));
        QL_REQUIRE(months != 0,
                   "invalid reference period for Act/365 Canadian; "
                   "must be longer than a month");
        auto frequency = Integer(12 / months);
        QL_REQUIRE(frequency != 0,
                   "invalid reference period for Act/365 Canadian; "
                   "must not be longer than a year");

        if (dcs < Integer(365/frequency))
            return dcs/365.0;

        return 1./frequency - (dcc-dcs)/365.0;

    }

    Date::serial_type Actual365Fixed::NL_Impl::dayCount(const Date& d1,
                                                        const Date& d2) const {

        static const Integer MonthOffset[] = {
            0,  31,  59,  90, 120, 151,  // Jan - Jun
            181, 212, 243, 273, 304, 334   // Jun - Dec
        };

        Date::serial_type s1 = d1.dayOfMonth()
                             + MonthOffset[d1.month()-1] + (d1.year() * 365);
        Date::serial_type s2 = d2.dayOfMonth()
                             + MonthOffset[d2.month()-1] + (d2.year() * 365);

        if (d1.month() == Feb && d1.dayOfMonth() == 29) {
            --s1;
        }

        if (d2.month() == Feb && d2.dayOfMonth() == 29) {
            --s2;
        }

        return s2 - s1;
    }

    Time Actual365Fixed::NL_Impl::yearFraction(const Date& d1,
                                               const Date& d2,
                                               const Date& d3,
                                               const Date& d4) const {
        return dayCount(d1, d2)/365.0;
    }

}

File: quantlib/time/daycounters/actual365fixed.py
from __future__ import annotations

from enum import IntEnum

from quantlib.time.date import Date, daysBetween
from quantlib.time.daycounter import DayCounter
from quantlib.time.month import Month


class Actual365Fixed(DayCounter):
    class Convention(IntEnum):
        Standard = 0
        Canadian = 1
        NoLeap = 2

    class _Impl(DayCounter.Impl):
        def name(self) -> str:
            return "Actual/365 (Fixed)"

        def yearFraction(self, d1: Date, d2: Date, refPeriodStart: Date, refPeriodEnd: Date) -> float:
            return daysBetween(d1, d2) / 365.0

    class _CA_Impl(DayCounter.Impl):
        def name(self) -> str:
            return "Actual/365 (Fixed) Canadian Bond"

        def yearFraction(self, d1: Date, d2: Date, refPeriodStart: Date, refPeriodEnd: Date) -> float:
            if d1 == d2:
                return 0.0
            if not refPeriodStart or not refPeriodEnd:
                raise ValueError("invalid refPeriodStart or refPeriodEnd")
            dcs = daysBetween(d1, d2)
            dcc = daysBetween(refPeriodStart, refPeriodEnd)
            months = round(12 * dcc / 365)
            if months == 0:
                raise ValueError("invalid reference period for Act/365 Canadian; must be longer than a month")
            frequency = 12 // months
            if frequency == 0:
                raise ValueError("invalid reference period for Act/365 Canadian; must not be longer than a year")
            if dcs < 365 / frequency:
                return dcs / 365.0
            return 1.0 / frequency - (dcc - dcs) / 365.0

    class _NL_Impl(DayCounter.Impl):
        _MONTH_OFFSET = [0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334]

        def name(self) -> str:
            return "Actual/365 (No Leap)"

        def dayCount(self, d1: Date, d2: Date) -> int:
            s1 = d1.dayOfMonth() + self._MONTH_OFFSET[int(d1.month()) - 1] + d1.year() * 365
            s2 = d2.dayOfMonth() + self._MONTH_OFFSET[int(d2.month()) - 1] + d2.year() * 365
            if d1.month() == Month.February and d1.dayOfMonth() == 29:
                s1 -= 1
            if d2.month() == Month.February and d2.dayOfMonth() == 29:
                s2 -= 1
            return s2 - s1

        def yearFraction(self, d1: Date, d2: Date, refPeriodStart: Date, refPeriodEnd: Date) -> float:
            return self.dayCount(d1, d2) / 365.0

    def __init__(self, convention: Actual365Fixed.Convention = Convention.Standard):
        if convention == Actual365Fixed.Convention.Standard:
            impl: DayCounter.Impl = Actual365Fixed._Impl()
        elif convention == Actual365Fixed.Convention.Canadian:
            impl = Actual365Fixed._CA_Impl()
        elif convention == Actual365Fixed.Convention.NoLeap:
            impl = Actual365Fixed._NL_Impl()
        else:
            raise ValueError(f"unknown Actual/365 (Fixed) convention: {convention}")
        super().__init__(impl)
